lib/Service/TextBlockService.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * SPDX-FileCopyrightText: 2025 Nextcloud GmbH and Nextcloud contributors
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+namespace OCA\Mail\Service;
+
+use Html2Text\Html2Text;
+use OCA\Mail\Db\TextBlock;
+use OCA\Mail\Db\TextBlockMapper;
+use OCA\Mail\Db\TextBlockShare;
+use OCA\Mail\Db\TextBlockShareMapper;
+use OCA\Mail\Exception\ShareeAlreadyExistsException;
+use OCA\Mail\Exception\UserNotFoundException;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\IGroupManager;
+use OCP\IUserManager;
+
+class TextBlockService {
+
+	public function __construct(
+		private TextBlockMapper $textBlockMapper,
+		private TextBlockShareMapper $textBlockShareMapper,
+		private IGroupManager $groupManager,
+		private IUserManager $userManager,
+	) {
+	}
+
+	/**
+	 * @param string $userId
+	 * @return TextBlock[]
+	 */
+	public function findAll(string $userId): array {
+		return $this->textBlockMapper->findAll($userId);
+	}
+
+	/**
+	 * @throws UserNotFoundException
+	 */
+	public function findAllSharedWithMe(string $userId): array {
+		$user = $this->userManager->get($userId);
+		if ($user === null) {
+			throw new UserNotFoundException();
+		}
+		$groups = $this->groupManager->getUserGroupIds($user);
+		return $this->textBlockMapper->findSharedWithMe($userId, $groups);
+	}
+	/**
+	 * @throws DoesNotExistException
+	 */
+	public function find(int $textBlockId, string $userId): ?TextBlock {
+		return $this->textBlockMapper->find($textBlockId, $userId);
+	}
+
+	public function create(string $userId, string $title, string $content): TextBlock {
+		$textBlock = new TextBlock();
+		$textBlock->setContent($content);
+		$textBlock->setOwner($userId);
+		$textBlock->setTitle($title);
+		$html = new Html2Text($content, ['do_links' => 'none','alt_image' => 'hide']);
+		$preview = trim($html->getText());
+		$textBlock->setPreview(substr($preview, 0, 50));
+		return $this->textBlockMapper->insert($textBlock);
+	}
+
+	/**
+	 * @throws DoesNotExistException
+	 */
+	public function update(TextBlock $textBlock, string $userId, string $title, string $content): TextBlock {
+		$textBlock->setContent($content);
+		$textBlock->setTitle($title);
+		$html = new Html2Text($content, ['do_links' => 'none','alt_image' => 'hide']);
+		$preview = trim($html->getText());
+		$textBlock->setPreview(substr($preview, 0, 300));
+		return $this->textBlockMapper->update($textBlock);
+	}
+
+	/**
+	 * @throws DoesNotExistException
+	 */
+	public function delete(int $textBlockId, string $userId): void {
+		$textBlock = $this->textBlockMapper->find($textBlockId, $userId);
+		$this->textBlockMapper->delete($textBlock);
```

Comment:
possible null value when the block doesn't exist

Reply:
https://github.com/nextcloud/server/blob/master/lib/public/AppFramework/Db/QBMapper.php#L374 `findEntity`doesn't return `null`, it  throws `DoesNotExistException` which I let propagate to the controller and catch there